pkg/controller/pcidevice/pcidevice_controller.go

```diff
@@ -49,42 +51,68 @@ func (h Handler) reconcilePCIDevices(nodename string) error {
 	for _, dev := range pci.Devices {
 		setOfRealPCIAddrs[dev.Address] = true
 		name := v1beta1.PCIDeviceNameForHostname(dev, nodename)
 		// Check if device is stored
-		_, err := h.client.Get(name, metav1.GetOptions{})
+		devCR, err := h.client.Get(name, metav1.GetOptions{})
 
 		// TODO use k8s.io/apimachinery/apierrors IsNotFound
 		if err != nil {
-			logrus.Errorf("Failed to get %s: %s\n", name, err)
-
-			// Create the PCIDevice CR if it doesn't exist
-			var pdToCreate v1beta1.PCIDevice = v1beta1.NewPCIDeviceForHostname(dev, nodename)
-			logrus.Infof("Creating PCI Device: %s\n", err)
-			pdToCreate.Labels = map[string]string{"nodename": nodename} // label
-			_, err := h.client.Create(&pdToCreate)
-			if err != nil {
-				logrus.Errorf("Failed to create PCI Device: %s\n", err)
+			if apierrors.IsNotFound(err) {
+				logrus.Infof("[PCIDeviceController] Device does %s does not exist: %s\n", name)
+
+				// Create the PCIDevice CR if it doesn't exist
+				var pdToCreate v1beta1.PCIDevice = v1beta1.NewPCIDeviceForHostname(dev, nodename)
+				logrus.Infof("Creating PCI Device: %s\n", err)
+				pdToCreate.Labels = commonLabels
+				_, err := h.client.Create(&pdToCreate)
+				if err != nil {
+					logrus.Errorf("[PCIDeviceController] Failed to create PCI Device: %v\n", err)
+					return err
+				}
+			} else {
+				logrus.Errorf("[PCIDeviceController] error fetching device %s: %v", name, err)
+				return err
 			}
+
 		}
-		// Update the stored device
-		devCR, err := h.client.Get(name, metav1.GetOptions{})
-		if err != nil {
-			logrus.Errorf("Failed to get %s: %s\n", name, err)
-			return err
-		}
+
 		devCopy := devCR.DeepCopy()
+		// Update only modifies the status, no need to update the main object
 		devCopy.Status.Update(dev, nodename) // update the in-memory CR with the current PCI info
-		_, err = h.client.Update(devCopy)
-		if err != nil {
-			logrus.Errorf("[PCIDeviceController] Failed to update resource: %s\n", err)
-		}
 		_, err = h.client.UpdateStatus(devCopy)
 		if err != nil {
-			logrus.Errorf("[PCIDeviceController] Failed to update status sub-resource: %s\n", err)
+			logrus.Errorf("[PCIDeviceController] Failed to update status sub-resource: %v\n", err)
+			return err
 		}
 	}
+
+	// remove non-existent devices
+	selector := labels.SelectorFromValidatedSet(commonLabels)
+
+	pdList, err := h.client.List(metav1.ListOptions{LabelSelector: selector.String()})
+	if err != nil {
+		logrus.Errorf("[PCIDeviceController] error listing devices for node %s: %v", nodename, err)
+		return err
+	}
+
+	var deleteList []v1beta1.PCIDevice
+
+	for _, v := range pdList.Items {
+		if ok := setOfRealPCIAddrs[v.Status.Address]; !ok {
+			deleteList = append(deleteList, v)
+		}
+	}
+
+	for _, v := range deleteList {
+		if err := h.client.Delete(v.Name, &metav1.DeleteOptions{}); err != nil {
+			logrus.Errorf("[PCIDeviceController] Faield to delete non existent device: %s on node %s", v.Name, v.Status.NodeName)
+			return err
```

Comment:
Is it too aggressive that we need to break down here when any delete operation fails?

This section does the work named `// remove non-existent devices`.
If we ignore the delete operation fails, that we would get the non-existent device on the client?
Just curious, What problems would cause by non-existent devices?

Reply:
I would assume the object exists and is available in the harvester UI for pcidevice claims and passthrough when in reality the backing physical device is now missing.

Reply:
OK, that makes sense to me. Thanks for explaining.